### ISSUE TITLE
bootstrap option to exclude curl/ssl

### ISSUE THREAD

**Issue by TurkeyMan** [premake/premake-core#529]
Can have Bootstrap.mak support an option to remove curl/ssl from the build?
I have to edit it each time.

**Comment by starkos** [comment]
I'd personally rather see this makefile kept as simple as possible, and encourage people to use Premake to configure the build instead in most cases. Is there a situation where it is not possible to bootstrap an initial Premake binary because of those libraries? If so, maybe we should consider dropping them from the bootstrap build, and require people to build a second binary via Premake if they need those features?

**Comment by TurkeyMan** [comment]
Bootstrap.mak already kinda does that. It doesn't seem to just build that bootstrap binary anymore, the script then runs the bootstrap binary to generate a gmake file and then builds a proper binary. It's the second step that includes curl/ssl:
```make
	mkdir -p build/bootstrap
	# builds the bootstrap binary
	$(CC) -o build/bootstrap/premake_bootstrap -DPREMAKE_NO_BUILTIN_SCRIPTS -DLUA_USE_POSIX -DLUA_USE_DLOPEN -I"$(LUA_DIR)" $? -lm -ldl -lrt
	./build/bootstrap/premake_bootstrap embed

	# generate a gmake file using the bootstrap (this is where I hack in --no-curl)
	./build/bootstrap/premake_bootstrap --to=build/bootstrap gmake
	# build a binary normally
	$(MAKE) -C build/bootstrap -j`getconf _NPROCESSORS_ONLN`
```
Your comment above kinda sounds like you feel the second step is beyond the scope of the bootstrap script?

**Comment by tvandijck** [comment]
I'm sure we could have an option that gets passed into the premake-bootstrap binary...

That said, the effort towards a module system that fetches modules from an online repository would then also be disabled. Since that will require http and likely https access.. I think the bigger question is, how do we solve the dependency on ssl in a way that allows premake to move around more freely between distributions..

In other words... instead of patching the problem with a band-aid, why don't we actually fix the problem?

**Comment by starkos** [comment]
> Your comment above kinda sounds like you feel the second step is beyond the scope of the bootstrap script?

I'd have to have another look, but I thought the bootstrap was building a barebones executable, and the second step built the full-featured version. I think that's reasonable. You should then be able to use that full-featured build to create a customized build, if you like. Is there a reason that doesn't work for you?

Shorter version: I think the bootstrap should do just enough to let someone with a bare repository get up and running. After that, you should switch over to using Premake itself.

> instead of patching the problem with a band-aid, why don't we actually fix the problem?

I have to admit, I'm not actually sure what the problem is yet.

**Comment by tvandijck** [comment]
Well, as presented, @TurkeyMan ran into the issue that if he build premake on one box, that version cannot be used on another box, due to the dependency on libssl, when you build using the bootstrap. This is mostly due to the fact that he builds the executable on 'ubuntu' (for example) and then tries to execute that binary on 'centos5' (just another example). These linux distributions come with different versions of libssl, and hence the dependency doesn't resolve because linux solved that in an absolutely stupid way.

There is several ways to fix it...
- Link against ssl in a static way (don't use .so). 
     - I actually have a PR locally with premake using mbedtls, for exactly the same reasons.
     - mbedtls is also about 100kb, instead of 5MB for openssl.
- Don't link against ssl at all.. this will disable https access however.

The downside to mbedtls is that we'd have to add the source for it to the premake repository, while not big, it's yet another dependency embedded in the codebase, all just for those two api's (http.download & http.get) It's a bummer "modules" can't have C/C++ in them.

**Comment by TurkeyMan** [comment]
I think we could also dynamically link to the less-specific version, ie, `libssl.so.1` (instead of `libssl.so.1.0.0` in my case). I'm pretty sure that's the whole reason those less-specific symlinks always exist.
It seems that the linker by default links to the most-specific version... which I find surprising.

I'm not familiar with `mbedtls`, but if it's small, I agree a static-linked alternative would better suit the premake ethos.

> The downside to mbedtls is [...]

Why do we keep putting OSS C libs in our source-tree? Shouldn't we just submodule their actual repo?

**Comment by tvandijck** [comment]
> Why do we keep putting OSS C libs in our source-tree? Shouldn't we just submodule their actual repo?

Yeah, I was thinking the same, we do however have to 'add' the premake5.lua script... and for curl we need autoconf, which we don't have in premake. So we have to actually modify the code itself.

for zlib and mbedtls however that would totally be an option.

**Comment by starkos** [comment]
> Yeah, I was thinking the same, we do however have to 'add' the premake5.lua script... and for curl we need autoconf, which we don't have in premake. So we have to actually modify the code itself.

The premake5.lua script can live outside of the subproject folder easily enough, but curl is still an issue. Though…do we actually modify the code, or do we just insert our own header?

That said, I'd be curious to know how we're getting linked to such a specific version. Premake should have a way to link more generically for cases like these?

**Comment by tvandijck** [comment]
We added:
```c

#ifdef __linux__
#  include "config-linux.h"
#endif

#ifdef __APPLE__ && __MACH__
#  include "config-osx.h"
#endif

```
to curl_setup.h and we then added those two files.

I don't think we modified libzip or zlib in anyway... and in our local version, we have no changes to mbedtls either... so for those three it would be easy... even 'lua' could probably be fetched that way, we made no changes to our 5.3 here, but I'm not sure if you did for your 5.1 version.

**Comment by tvandijck** [comment]
Also,
curl is here: https://github.com/curl/curl
zlib is here: https://github.com/madler/zlib
libzip is here: https://github.com/nih-at/libzip
mbedtls is here: https://github.com/ARMmbed/mbedtls
lua is here: https://github.com/lua/lua

So they are all on github, and can hence be referenced as submodules.
We'd have to reference specific tags/releases however.

**Comment by ratzlaff** [comment]
So I am sure I have missed the original discussion around including these libraries but...

Why are these libraries in the default build? It is not obvious to me how curl and zlib help me generate a build system for linux/mac/windows

Having used cmake, it is completely **irritating** how it does **absolutely nothing** when it is unable to discover all of its dependencies. It is extremely nice how premake makes you a build system just exactly how you tell it. 

I personally do not want premake to ever download anything from anywhere. I do not see why this is considered a good thing. If the environment is missing files necessary to the generation of a project, I would consider that a problem with how the source is controlled and retrieved (with svn:externals or git submodules), not something the build system generator would handle.

That being said, I believe the project is safe in all of your capable hands and trust you have already considered all these issues. Thanks again for an excellent product!

**Comment by tvandijck** [comment]
> I personally do not want premake to ever download anything from anywhere. 

Honestly, if you've ever worked with Rust, Nuget in C#, or many other languages that have integrated package management systems, you'd probably not be saying this...

For me the power of having premake be able to download zlib from somewhere, and add it to my project is extremely useful. Just having a premake5.lua script that does this:

```lua
solution "A"
   import('zlib', '1.2.8');
   project 'b'
        uses 'zlib'
        files '*.cpp'
```

and I'm done, it's awesome. I don't have to go manually download zlib and put it in my project, setup a premake5 file for it, etc. etc..  So yeah, we differ in opinion.. I want premake to download EVERYTHING I use, except for the project I'm working on myself. But I should not have to worry about 3rd party dependencies...

Anyway, the reason it go added, is because premake doesn't allow modules with C/C++ in it. If it did, the whole curl/zlib/libzip integration could simply be done in a module... I wish we could solve that instead, and get vanilla premake a little less dependent on external code.

**Comment by ratzlaff** [comment]
> premake doesn't allow modules with C/C++ in it

Thats the answer I was looking for, thanks!

**Comment by starkos** [comment]
>  I wish we could solve that instead, and get vanilla premake a little less dependent on external code.

Me too! The big issue is splitting the Lua runtime out into a shared library, since both Premake and the module need to link against (right? I'm not imagining that?). With the catch that the shared library would need Premake's modifications to the Lua runtime…or we need to find a way to get the same behavior without modifying the library.

**Comment by tvandijck** [comment]
I think it would be easier to just make our own "marshaling" layer...

For example, if a 'module' had both lua and a dll, the dll could simply export a 'RegisterPremakeModule' method that we give it some kind of API to call into, and we can abstract the LuaState in a similar way.

If we can keep the "LuaState" completely outside of the module, and all interaction with lua itself goes 'through' a thing that the executable provides to the module, then we could easily do this. It creates a layer of abstraction, and hence a slight performance hit, but for something like http.download("...") I think a few cycles of overhead is hardly going to be an issue... Besides, it's not like lua is really fast to begin with.

All we really need for a module to work that way is provide module writers with a 'header file' that contains "our" API... there isn't even any code to link against. I'll prototype something simple as a start.

**Comment by tvandijck** [comment]
> would we also need to wrap the entire Lua C API?

Probably not all of it... really all that you need to provide a method to lua is for that method to be able to get values on and off the lua stack, and maybe create a lua table and populate it.

> a.

That would be correct. Premake modules would have their own api, and hence be unique to premake.

> b.

Yeah, there is good reasons to keep it the way it currently is if we wanted to integrate a package managment system... I've gone back and forth about using a separate tool, vs integration. There certainly is pro's and con's. Even if you did make an external tool for this however, premake itself would have to 'know' how it worked, so it can properly discover downloaded modules/packages.

For example, in our current system here at Blizzard, we download 'packages' into a package cache, which for my machine lives in C:\package_cache... completely outside the source tree of our game. Multiple branches share the same package source, and downloading packages happens rarely, because I already have most everything from a previous build. 

If a tool became external to premake, premake would still need to know the concept of packages, regardless of what/who downloaded it. That said, as an external tool, other systems may also be able to use it, msbuild, scons, whatever... so there is definitely ups and downs.

**Comment by starkos** [comment]
Been thinking about this…regardless what we decide re: C code modules, I think we should continue to building the HTTP and .zip functionality into the core build.

Premake's primary purpose is to generate project files, but part of that is—and I think should be—general purpose, cross-platform build configuration. Having those features baked in enables the creating of portable setup and deployment scripts, GitHub scripting, and also happens to enable the module and package systems we've been discussing. So I think it is totally worth the extra work to get it all working smoothly in core, rather than trying to ship it as an add-on.

My $0.02.

**Comment by PaulVisscher** [comment]
Regarding the package management, we at Zefiros Software have been working on a solution http://zpm.zefiros.eu/, whilst everything is functional it is still in beta. It could be used as a base solution.

**Comment by samsinsane** [comment]
Since this issue was based on our usage of OpenSSL, I'm going to close this since we've removed it. Please reopen if the issue is still occurring.
